subprojects/dependency-management/src/test/groovy/org/gradle/internal/resolve/caching/ComponentMetadataRuleExecutorTest.groovy

```diff
@@ -0,0 +1,242 @@
+        detailsToResult = Mock()
+        onCacheMiss = Mock()
+        executor = new ComponentMetadataRuleExecutor(cacheRepository, cacheDecoratorFactory, valueSnapshotter, timeProvider, serializer)
+    }
+
+    // Tests --refresh-dependencies behavior
+    @Unroll("Cache expiry check age=#age, refresh = #mustRefresh - #scenario - #ruleClass")
+    def "expires entry when cache policy tells us to"() {
+        def id = DefaultModuleVersionIdentifier.newId('org', 'foo', '1.0')
+        def hashValue = Mock(HashValue)
+        def key = Mock(ModuleComponentResolveMetadata)
+        def inputsSnapshot = new StringValueSnapshot("1")
+        def cachedResult = Mock(ModuleComponentResolveMetadata)
+        Multimap<String, ImplicitInputRecord<?, ?>> implicits = HashMultimap.create()
+        def record = Mock(ImplicitInputRecord)
+        if (expired) {
+            implicits.put('SomeService', record)
+        }
+        def cachedEntry = new CrossBuildCachingRuleExecutor.CachedEntry<ComponentMetadata>(0, implicits, cachedResult)
+        def ruleServices = [:]
+        def someService = Mock(SomeService)
+        if (ruleClass == TestSupplierWithService) {
+            ruleServices['SomeService'] = someService
+        }
+        def reexecute = mustRefresh || expired
+
+        when:
+        withRule(ruleClass, ruleServices)
+        execute(key)
+
+        then:
+        1 * key.contentHash >> hashValue
+        1 * hashValue.asBigInteger() >> new BigInteger("42")
+        1 * valueSnapshotter.snapshot(_) >> inputsSnapshot
+        1 * store.get(inputsSnapshot) >> cachedEntry
+        if (expired) {
+            // should check that the recorded service call returns the same value
+            1 * record.getInput() >> '124'
+            1 * record.getOutput() >> HashCode.fromInt(10000)
+            1 * cachedResult.isChanging() >> changing
+            1 * cachedResult.getModuleVersionId() >> id
+            1 * cachePolicy.mustRefreshModule({ it.id == id }, 0, changing) >> false
+            // we make it return false, this should invalidate the cache
+            1 * someService.isUpToDate('124', HashCode.fromInt(10000)) >> false
+        } else {
+            1 * cachedResult.isChanging() >> changing
+            1 * cachedResult.getModuleVersionId() >> id
+            1 * cachePolicy.mustRefreshModule({ it.id == id }, 0, changing) >> mustRefresh
+        }
+        if (reexecute) {
+            def details = Mock(ComponentMetadataContext)
+            if (ruleClass == TestSupplierWithService) {
+                // indicates that the service will be called
+                1 * someService.withImplicitInputRecorder(_) >> someService
+            }
+            1 * onCacheMiss.transform(key) >> details
+            1 * detailsToResult.transform(details) >> Mock(ModuleComponentResolveMetadata)
+            1 * store.put(inputsSnapshot, _)
+        }
+        // This shows a limitation of the current implementation:
```

Comment:
:x: This is no longer true, probably copy and paste from outdated branch.

Reply:
Removed in both places